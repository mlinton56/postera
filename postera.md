VERSION 0.3.0
Modules for developing and connecting software components
README
## postera

The postera package is a collection of the following TypeScript modules:

<ul>
<li><a href="doc/kstore.md">kstore</a> &mdash;
Use a key-value store (e.g., Redis) to share data</li>
<li><a href="doc/logsearch.md">kstore</a> &mdash;
Search logs (e.g., Papertrail)</li>
<li><a href="doc/qio.md">qio</a> &mdash;
Use queues to send and receive messages</li>
<li><a href="doc/notifier.md">notifier</a> &mdash;
Base class for posting notifications</li>
<li><a href="doc/reqm.md">reqm</a> &mdash;
Perform and track HTTP requests using promises.</li>
<li><a href="doc/slogger.md">slogger</a> &mdash;
Shared logging object</li>
<li><a href="doc/tshell.md">tshell</a> &mdash;
Run shell commands</li>
</ul>
EOF
