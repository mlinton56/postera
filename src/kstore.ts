/**
 * Module that provides a key-value store interface with an implementation
 * that uses a Redis server.
 */

import logger from './slogger'

const redis = require('redis')

export interface KeyValueStore {
    impl: any

    value(key: string): Promise<any>
    strValue(key: string): Promise<string>

    valueMod(key: string, value: any): Promise<void>
    strValueMod(key: string, value: string): Promise<void>

    valueIncr(key: string): Promise<number>
    valueDel(keys: string[]): Promise<number>
    valueDelAfter(key: string, seconds: number): Promise<boolean>

    map<T extends object>(key: string): Promise<T>
    mapMod<T extends object>(key: string, changes: T): Promise<void>
    mapDel(key: string, i: string[]): Promise<number>
    mapSize(key: string): Promise<number>
    mapItem(key: string, i: string): Promise<any>
    mapItemMod(key: string, i: string, value: any): Promise<boolean>
    mapItemIncr(key: string, i: string): Promise<number>

    strMap<T extends object>(key: string): Promise<T>
    strMapMod<T extends object>(key: string, changes: T): Promise<void>
    strMapItem(key: string, i: string): Promise<string>
    strMapItemMod(key: string, i: string, value: string): Promise<boolean>
}

export function redisStore(config): KeyValueStore {
    const store = new RedisStore()
    store.impl = redis.createClient(config)
    logger.info('Redis server is ' + config.host)
    return store
}


export type Result = (err, res) => void

export class RedisStore implements KeyValueStore {

    impl: any

    value(key: string): Promise<any> {
        return new Promise<any>((resolve, reject) =>
            this.impl.get(key, parseResult(resolve, reject))
        )
    }

    strValue(key: string): Promise<string> {
        return new Promise<any>((resolve, reject) =>
            this.impl.get(key, result(resolve, reject))
        )
    }

    valueMod(key: string, value: any): Promise<void> {
        return new Promise<void>((resolve, reject) =>
            this.impl.set(key, JSON.stringify(value), voidResult(resolve, reject))
        )
    }

    strValueMod(key: string, value: string): Promise<void> {
        return new Promise<void>((resolve, reject) =>
            this.impl.set(key, value, voidResult(resolve, reject))
        )
    }

    valueIncr(key: string): Promise<number> {
        return new Promise<number>((resolve, reject) =>
            this.impl.incr(key, parseResult(resolve, reject))
        )
    }

    valueDel(keys: string[]): Promise<number> {
        return new Promise<number>((resolve, reject) =>
            this.impl.del(...keys, parseResult(resolve, reject))
        )
    }

    valueDelAfter(key: string, seconds: number): Promise<boolean> {
        return new Promise<boolean>((resolve, reject) =>
            this.impl.expire(key, seconds, (err, res) => {
                err ? reject(err) : resolve(res == 1)
            })
        )
    }

    map(key: string): Promise<any> {
        return new Promise<any>((resolve, reject) => {
            this.impl.hgetall(key, (err, res) => {
                if (err) {
                    reject(err)
                } else {
                    try {
                        const items = {}
                        if (res) {
                            for (let k of Object.keys(res)) {
                                items[k] = JSON.parse(res[k])
                            }
                        }
                        resolve(items)
                    } catch (e) {
                        reject(e)
                    }
                }
            })
        })
    }

    mapMod<T extends object>(key: string, changes: T): Promise<void> {
        const fields = Object.keys(changes)
        if (fields.length === 1) {
            const name = fields[0]
            const value = JSON.stringify(changes[name])
            return new Promise<void>((resolve, reject) =>
                this.impl.hset(key, name, value, voidResult(resolve, reject))
            )
        } else {
            let args = []
            for (let f of fields) {
                args.push(f)
                args.push(JSON.stringify(changes[f]))
            }

            return new Promise<void>((resolve, reject) =>
                this.impl.hmset(key, ...args, voidResult(resolve, reject))
            )
        }
    }

    mapDel(key: string, i: string[]): Promise<number> {
        return new Promise<any>((resolve, reject) =>
            this.impl.hdel(key, ...i, parseResult(resolve, reject))
        )
    }

    mapSize(key: string): Promise<number> {
        return new Promise<number>((resolve, reject) =>
            this.impl.hlen(key, parseResult(resolve, reject))
        )
    }

    mapItem(key: string, i: string): Promise<any> {
        return new Promise<any>((resolve, reject) =>
            this.impl.get(key, i, parseResult(resolve, reject))
        )
    }

    mapItemMod(key: string, i: string, value: any): Promise<boolean> {
        return new Promise<boolean>((resolve, reject) =>
            this.impl.hset(key, i, JSON.stringify(value), (err, res) => 
                err ? reject(err) : resolve(res == 1)
            )
        )
    }

    mapItemIncr(key: string, i: string): Promise<number> {
        return new Promise<number>((resolve, reject) =>
            this.impl.hincrby(key, i, 1, parseResult(resolve, reject))
        )
    }

    strMap(key: string): Promise<any> {
        return new Promise<any>((resolve, reject) =>
            this.impl.hgetall(key, result(resolve, reject))
        )
    }

    strMapMod<T extends object>(key: string, changes: T): Promise<void> {
        const fields = Object.keys(changes)
        if (fields.length === 1) {
            const name = fields[0]
            const value = changes[name]
            return new Promise<void>((resolve, reject) =>
                this.impl.hset(key, name, value, voidResult(resolve, reject))
            )
        } else {
            let args = []
            for (let f of fields) {
                args.push(f)
                args.push(changes[f])
            }

            return new Promise<void>((resolve, reject) =>
                this.impl.hmset(key, ...args, voidResult(resolve, reject))
            )
        }
    }

    strMapItem(key: string, i: string): Promise<string> {
        return new Promise<string>((resolve, reject) =>
            this.impl.hget(key, i, result(resolve, reject))
        )
    }

    strMapItemMod(key: string, i: string, value: string): Promise<boolean> {
        return new Promise<boolean>((resolve, reject) =>
            this.impl.hset(key, i, value, (err, res) => 
                err ? reject(err) : resolve(res == 1)
            )
        )
    }

}

function result(resolve, reject): Result {
    return (err, res) => err ? reject(err) : resolve(res)
}

function parseResult(resolve, reject): Result {
    return (err, res) => {
        if (err) {
            reject(err)
        } else {
            try {
                resolve(res ? JSON.parse(res) : null)
            } catch (e) {
                reject(e instanceof Error ? <Error>e: new Error(e.toString()))
            }
        }
    }
}

function voidResult(resolve, reject): Result {
    return (err, res) => err ? reject(err) : resolve()
}
