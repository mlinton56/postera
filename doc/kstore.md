## kstore

The kstore module provides common methods to access a shared key-value store,
with an implementation for Redis. The methods are promise-based
to simplify coding synchronous use cases.

The method interfaces support associating strings or objects with a string key,
as well as maps of strings or objects accessed by an item identifier (string).

    import * as kstore from 'postera/kstore'

    // Access a key-value store with a given configuration.
    const store = kstore.redisStore({host: 'localhost'})

    // Modify the value associated with a key.
    await store.valueMod('object', {a: 3, b: 4})

    // Access the associated value, which is {a: 3, b: 4}.
    const obj = await store.value('object')

    // Delete an association
    await store.valueDel('object')

Note that values are normally converted to and from strings with
JSON.stringify and JSON.parse. To avoid conversion use string values:

    await store.strValueMod('object', 'this is a string')

#### Maps

A map is like another value (it can be deleted with valueDel, for example)
but allows access by item without retrieving the full map.

    // Associate a map.
    await store.mapMod('map', {a: 3, b: 4})

    //
    // Access an item without getting the entire map. In this case
    // the return value is 3.
    //
    await store.mapItem('map', 'a')

    // Modify an item within a map.
    if (await store.mapItemMod('map', 'c', 5)) {
        // Returns true if the item is new to the map.
    }

    // Access the entire map at once--this returns {a: 3, b: 4}.
    await store.map('map')

As with values, map items are normally converted to and from strings
with JSON.stringify and JSON.parse. To avoid conversion use string maps:

    await store.strMapMod('map', {'a': 'a string', b: 'b string'})
    await store.strMapItem('map', 'a')
    await store.strMapItemMod('map', 'c', 'c string')
    await store.strMap('map')
