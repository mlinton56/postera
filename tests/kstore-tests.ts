/**
 * Test case for postera/kstore.
 *
 * These tests are a bit convoluted because of the specific use cases in mind.
 * Hopefully that doesn't mean the KeyValueStore interface is also convoluted!
 */

import * from 'postera/kstore'
import logger from 'postera/slogger'

export interface BatchInfo {
    user: string
    keys: string[]
}

const store = kstore.RedisStore({host: process.env.REDIS_HOST || 'localhost'})

async function main() {
    const batchId = (await store.valueIncr('batchId')).toString()
    await addBatch(batchId, 'user', {
        'first': 'This is the first item',
        'second': 'This is the second item'
    })

    await status('user', '100')
    await status('user', batchId)
    await status('other', batchId)

    await addResult(batchId, 'first', 'This is the first result')
    await status('user', batchId)
    await addResult(batchId, 'second', 'This is the second result')
    await status('user', batchId)
}

async function addBatch(batchId: string, user: string, batch) {
    const keys = Object.keys(batch)
    await store.mapMod(batchId, {
        info: {user: user, keys: keys},
        length: keys.length
    })
}

async function addResult(b: string, item: string, result: string) {
    if (await store.mapItemMod(b + '/', item, result)) {
        if (await store.mapItem(b, 'length') === await store.mapSize(b + '/')) {
            const info = await store.mapItem(b, 'info')
            const results = await store.map(b + '/')
            if (info && results) {
                console.log(results)
            } else {
                logger.error('Cannot get info or results for ' + b)
            }
        }
    }
}

async function status(user: string, b: string) {
    const info = await store.mapItem(b, 'info')
    if (!info) {
        logger.error('Unknown batch ' + b)
    } else if (info.user !== user) {
        logger.error('Wrong user')
    } else {
        console.log(await store.map(b + '/'))
    }
}

main()
