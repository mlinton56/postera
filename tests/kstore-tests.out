Redis server is localhost
batchId = 1
added batch 1
getting status of batch 100 for user
Unknown batch 100
getting status of batch 1 for user
{"first":"This is the first result","second":"This is the second result"}
getting status of batch 1 for other
Wrong user
getting status of batch 1 for user
{"first":"This is the first result","second":"This is the second result"}
getting status of batch 1 for user
{"first":"This is the first result","second":"This is the second result"}
Creating subscriber 1
Creating subscriber 2
Publishing first message
2: "This is the first message"
1: "This is the first message"
Publishing second message
Deleting subscriber 1
2: "This is the second message"
Publishing third message
2: "This is the third message"
Deleting subscriber 2
Publishing fourth message
